schunk_gripper_driver/schunk_gripper_driver/tests/test_logging.py

```diff
@@ -39,3 +46,183 @@ def test_driver_doesnt_fill_disk_space_by_default(log_monitor, lifecycle_interfa
+
+
+@skip_without_gripper
+def test_driver_rejects_invalid_log_level(driver):
+    client = rclpy.create_node("test_bad_log_level")
+    set_params_client = client.create_client(
+        SetParameters, "/schunk/driver/set_parameters"
+    )
+    assert set_params_client.wait_for_service(timeout_sec=2)
+    get_params_client = client.create_client(
+        GetParameters, "/schunk/driver/get_parameters"
+    )
+    assert get_params_client.wait_for_service(timeout_sec=2)
+
+    # Attempt to set an invalid log level
+    future = set_params_client.call_async(
+        SetParameters.Request(
+            parameters=[
+                Parameter(
+                    name="log_level",
+                    value=ParameterValue(
+                        type=ParameterType.PARAMETER_STRING,
+                        string_value="BAD_LOG_LEVEL",
+                    ),
+                )
+            ]
+        )
+    )
+    rclpy.spin_until_future_complete(client, future, timeout_sec=2)
+    assert future.result().results[0].successful is False
+    # Verify the log level was not set to the invalid value
+    future = get_params_client.call_async(GetParameters.Request(names=["log_level"]))
+    rclpy.spin_until_future_complete(client, future)
+    assert future.result().values[0].string_value != "BAD_LOG_LEVEL"
+
+
+@skip_without_gripper
+def test_driver_logs_correct_level(driver):
+    log_level = []
+
+    driver = Driver("test_driver_logs_correct_level")
+    executor = rclpy.executors.SingleThreadedExecutor()
+    executor.add_node(driver)
+
+    executor_thread = threading.Thread(target=executor.spin, daemon=True)
+    executor_thread.start()
+
+    client = rclpy.create_node("test_log_output_set_level")
+    set_params_client = client.create_client(
+        SetParameters, "/test_driver_logs_correct_level/set_parameters"
+    )
+
+    assert set_params_client.wait_for_service(timeout_sec=5), "Service not available"
+
+    future = set_params_client.call_async(
+        SetParameters.Request(
+            parameters=[
+                Parameter(
+                    name="log_level",
+                    value=ParameterValue(
+                        type=ParameterType.PARAMETER_STRING, string_value="ERROR"
+                    ),
+                )
+            ]
+        )
+    )
+
+    # Spin the client node until the future is completed
+    rclpy.spin_until_future_complete(client, future)
+    assert future.result().results[0].successful, "Failed to set log level"
+
+    subscription_client = Node("test_log_output")
+
+    def callback(data: Log):
+        log_level.append(data.level)
+
+    subscription = subscription_client.create_subscription(
+        Log,
+        "/rosout",
+        callback,
+        10,
+    )
+    if not subscription:
+        # use of subscription variable needed for checkstyle
+        raise RuntimeError("Failed to create subscription")
+
+    debug_num = 2
```

Comment:
This looks like too much code for what it does. Can we use a single loop?
Also, what's happening here?
Maybe you could add more high-level comments in this test what you are doing and why?